apps/drec-api/src/pods/issuer/issuer.service.ts

```diff
@@ -0,0 +1,164 @@
+        values(groupBy(ownerBasedGroup, deviceGroupRule2)).map(
+          async (countryBasedGroup: IDevice[], j) => {
+            const categorizedGroup: IDeviceGroup = {
+              id: 0,
+              name: `${ownerBasedGroup[i][deviceGroupRule1]}_${countryBasedGroup[j][deviceGroupRule2]}`,
+              organizationId: ownerBasedGroup[i].registrant_organisation_code,
+              devices: countryBasedGroup,
+            };
+            return await this.issueCertificateForGroup(
+              categorizedGroup,
+              startDate,
+              endDate,
+            );
+          },
+        );
+      }),
+    );
+  }
+
+  private async issueCertificateForGroup(
+    group: IDeviceGroup,
+    startDate: DateTime,
+    endDate: DateTime,
+  ): Promise<void> {
+    const readsFilter: FilterDTO = {
+      offset: 0,
+      limit: 1000,
+      start: startDate.toString(),
+      end: endDate.toString(),
+    };
+
+    const org = await this.organizationService.findOne(group.organizationId);
+    if (!org) {
+      throw new NotFoundException(
+        `No organization found with code ${group.organizationId}`,
+      );
+    }
+    const groupReads: number[] = [];
+    await Promise.all(
+      group.devices.map(async (device: IDevice) =>
+        groupReads.push(await this.getDeviceFullReads(device.id, readsFilter)),
+      ),
+    );
+    const totalReadValue = groupReads.reduce(
+      (accumulator, currentValue) => accumulator + currentValue,
+      0,
+    );
+    // Convert from W to kW
+    const totalReadValueKw = Math.round(totalReadValue * 10 ** -3);
+    const issuance: IIssueCommandParams<ICertificateMetadata> = {
+      deviceId: group.id.toString(), // groupID
+      energyValue: totalReadValueKw.toString(),
+      fromTime: new Date(startDate.toString()),
+      toTime: new Date(endDate.toString()),
+      toAddress: org.blockchainAccountAddress,
+      userId: org.blockchainAccountAddress,
+      metadata: {
+        deviceIds: group.devices.map((device: IDevice) => device.id),
+        id: group.id.toString(),
+      },
+    };
+    this.logger.log(
+      `Issuance: ${JSON.stringify(issuance)}, Group name: ${group.name}`,
+    );
+    return await this.issueCertificate(issuance);
+  }
+
+  private async getDeviceFullReads(
+    deviceId: number,
+    filter: FilterDTO,
+  ): Promise<number> {
+    const allReads = await this.baseReadsService.find(
+      deviceId.toString(),
+      filter,
+    );
+    return allReads.reduce(
+      (accumulator, currentValue) => accumulator + currentValue.value,
+      0,
+    );
+  }
+
+  private async issueCertificate(
+    reading: IIssueCommandParams<ICertificateMetadata>,
+  ): Promise<void> {
+    this.logger.log(`Issuing a certificate for reading`);
+    const issuedCertificate = await this.certificateService.issue(reading);
+    this.logger.log(`Issued a certificate with ID ${issuedCertificate.id}`);
```

Comment:
Not sure if these logs are required, since certificateService is already logging by itself